environment.go

```diff
@@ -57,9 +120,65 @@ func (e *PipelineEnvironment) exportEnvironment(path string) error {
 
 func (e *PipelineEnvironment) createTemplateParser() *template.Template {
 	fm := template.FuncMap{}
+	// {{ Key }}
+	// Required environment value, if not defined it will not be found as function and raise an error
 	for k, v := range e.Environment {
-		fm[k] = func() string { return *v }
+		// Convert string pointer to static string
+		fm[k] = func(v string) func() (string, error) {
+			return func() (string, error) {
+				return v, nil
+			}
+		}(*v)
+	}
+	// {{ Env "Key" ["Default"] }}
+	// Usable as optional environment variable, can provide default value if not defined
+	fm["Env"] = func(args ...interface{}) (string, error) {
+		if len(args) < 1 {
+			return "", errors.New(fmt.Sprintf("Env: missing argument(s). Need atleast 1 argument"))
+		}
+		if len(args) > 2 {
+			return "", errors.New(fmt.Sprintf("Env: too many arguments. Got %d want <=2", len(args)))
+		}
+		parts := make([]string, len(args))
+		for i, v := range args {
+			parts[i] = fmt.Sprint(v)
+		}
+		val, ok := e.Environment[parts[0]]
+		if !ok {
+			if len(parts) < 2 {
+				return "", errors.New(fmt.Sprintf("Env '%s' not defined, no fallback provided", parts[0]))
+			}
+			return parts[1], nil
+		}
+		return *val, nil
 	}
+	// {{ EnvDir "Key" ["Default"] }}
+	// Get Path from environment, converts to absolute path using filepath.Abs
+	fm["EnvDir"] = func(args ...interface{}) (string, error) {
+		if len(args) < 1 {
+			return "", errors.New(fmt.Sprintf("EnvDir: missing argument(s). Need atleast 1 argument"))
+		}
+		if len(args) > 2 {
+			return "", errors.New(fmt.Sprintf("EnvDir: too many arguments. Got %d want <=2", len(args)))
+		}
+		parts := make([]string, len(args))
+		for i, v := range args {
+			parts[i] = fmt.Sprint(v)
+		}
+		var path string
+		val, ok := e.Environment[parts[0]]
+		if ok {
+			path = *val
+		} else {
+			if len(parts) < 2 {
+				return "", errors.New(fmt.Sprintf("EnvDir '%s' not defined, no fallback provided", parts[0]))
+			}
+			path = parts[1]
+		}
+		return filepath.Abs(path)
+	}
+	// {{ TempDir ["optional" ["optional" ... ]] }}
+	// TempDir with the same arguments point to the same directory
```

Comment:
I don't understand what this comment is saying, "same arguments" as what? Which "same directory"?

Reply:
I try to improve the wording.